Rubberduck.Parsing/VBA/VBACodeStringParser.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using Antlr4.Runtime;
+using Antlr4.Runtime.Atn;
+using Antlr4.Runtime.Tree;
+using NLog;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Symbols.ParsingExceptions;
+
+namespace Rubberduck.Parsing.VBA
+{
+    /// <summary>
+    /// The class is meant to be used for where we need to do rewriting without actually
+    /// rewriting an existing module in a VBA project. The result is completely free-floating
+    /// and it is the caller's responsibility to persist the finalized results into one of
+    /// actual code. This is useful for creating a code preview for refactoring.
+    /// </summary>
+    public class VBACodeStringParser
+    {
+        public enum ParserMode
+        {
+            Default,
+            Sll,
+            Ll
+        }
+
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+        private readonly CommonTokenStream _tokenStream;
+        private readonly VBAParser _parser;
+
+        public TokenStreamRewriter Rewriter => new TokenStreamRewriter(_tokenStream);
+        public IParseTree ParseTree { get; }
+
+        /// <summary>
+        /// Parse a given string representing a module code. The code passed in must be
+        /// a valid module body, containing valid declarations or complete procedures. 
+        /// Code snippets are not valid. 
+        /// </summary>
+        /// <param name="moduleName">For logging purpose, provide descritpive name for where the virtual module is used</param>
+        /// <param name="moduleCodeString">A string containing valid module body.</param>
+        /// <param name="mode">Indicates what parser mode to use. By default we use SLL fallbacking to LL. When mode is explicitly specified, only that mode will be used with no fallback.</param>
+        public VBACodeStringParser(string moduleName, string moduleCodeString, ParserMode mode = ParserMode.Default) :
+            this(moduleName,
+                moduleCodeString)
+        {
+            ParseTree = mode == ParserMode.Default ? ParseInternal(moduleName) : ParseInternal(mode);
```

Comment:
I have to say that I do not really like that the parser parses upon construction. I would really prefer a design with an explicit `Parse` method that takes the code and maybe the module name and returns a value tuple `(IParseTree tree, ITokenStreamRewriter rewriter)`.

Reply:
Agreed. I updated accordingly. Note that we currently do not have an interface for the rewriter though we definitely should have one. I would prefer to deal with that deficiency in a separate PR, however.